io/reed_solomon_dagreader.go

```diff
@@ -0,0 +1,152 @@
+package io
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+
+	cid "github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	rs "github.com/klauspost/reedsolomon"
+)
+
+// reedSolomonDagReader reads a dag by concurrently merging N shards
+// in a N/M encoded UnixFS dag file.
+// Due to reed solomon requiring N shards to exist, we cannot perform
+// stream Read or Seek on this reader.
+// Everything will be pre-filled in memory before supporting DagReader
+// operations from a []byte reader.
+type reedSolomonDagReader struct {
+	*bytes.Reader // for Reader, Seeker, and WriteTo
+}
+
+type nodeBufIndex struct {
+	b   *bytes.Buffer
+	i   int
+	err error
+}
+
+// A ReedSolomonDagReader wraps M DagReaders and reads N (data) out of
+// M (data + parity) concurrently to decode the original file shards for
+// the returned DagReader to use.
+func NewReedSolomonDagReader(ctx context.Context, n ipld.Node, serv ipld.NodeGetter,
+	numData, numParity, size uint64) (DagReader, error) {
+	// TODO: Read data, parity and size from metadata when implemented
+
+	totalShards := int(numData + numParity)
+	if totalShards != len(n.Links()) {
+		return nil, fmt.Errorf("number of links under node [%d] does not match set data + parity [%d]",
+			len(n.Links()), totalShards)
+	}
+
+	// Grab at least N nodes then we are ready for re-construction
+	// Timeout is set at upper caller level through context.Context
+	nodeBufChan := make(chan nodeBufIndex)
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	for i, link := range n.Links() {
+		go func(ctx context.Context, shardCID cid.Cid, index int) {
+			node, err := serv.Get(ctx, shardCID)
+			if err != nil {
+				nodeBufChan <- nodeBufIndex{nil, index, err}
+				return
+			}
+			dr, err := NewDagReader(ctx, node, serv)
+			if err != nil {
+				nodeBufChan <- nodeBufIndex{nil, index, err}
+				return
+			}
+			var b bytes.Buffer
+			_, err = io.Copy(&b, dr)
+			if err != nil {
+				nodeBufChan <- nodeBufIndex{nil, index, err}
+				return
+			}
+			nodeBufChan <- nodeBufIndex{&b, index, nil}
+		}(ctxWithCancel, link.Cid, i)
+	}
+
+	// Context deadline is set so it should exit eventually
+	bufs := make([]*bytes.Buffer, totalShards)
+	valid := 0
+	for nbi := range nodeBufChan {
+		if nbi.err != nil {
+			continue
+		}
+		bufs[nbi.i] = nbi.b
+		valid += 1
+		if valid == int(numData) {
+			// No need to get more nodes
+			cancel()
+			break
+		}
+	}
+	if valid < int(numData) {
+		return nil, fmt.Errorf("unable to obtain at least [%d] shards to join original file", numData)
+	}
+
+	// Check if we already have everything
+	dataValid := true
+	for i := 0; i < int(numData); i++ {
+		if bufs[i] == nil {
+			dataValid = false
+			break
+		}
+	}
+
+	// Create rs stream
+	rss, err := rs.NewStreamC(int(numData), int(numParity), true, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// Reconstruct if missing some data shards
+	if !dataValid {
```

Comment:
Do we need extra validation other than read error case? Like checking a hash value to validate each shard? 
I am referring to https://github.com/klauspost/reedsolomon.  But at this layer it should be already done by a lower layer in IPFS.

Reply:
I am missing a Verify func. Will add it.

Reply:
Actually, for performance reasons, I cannot use Verify. Verify requires all shards to exist. But we are only getting the minimum 10 to quickly recover data. Computing all 30 would take longer.

For your concern, BTFS layer also guarantees file consistency.